### PR TITLE
Inhibit incompatible ARM upgrades

### PR DESCRIPTION
**Pull request by dkubek**
Due to an incompatibility of RHEL8 bootloader with newer versions of kernel on RHEL9 since version 9.5 the upgrade cannot be performed as the old bootloader is used to load the new versions of kernel during the upgrade.

TODO:
  - [x] add ~KCS~ docs link

JIRA: [41193](https://issues.redhat.com/browse/RHEL-41193), [52993](https://issues.redhat.com/browse/RHEL-52993)

### CODE REVIEW

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        reporting.Remediation(hint=(
+            'To upgrade to a RHEL version 9.5 and up, first perform an upgrade '
+            'to version 9.4 using the leapp `--target` option and then perform '
+            'a minor dnf update after the upgrade')),
+    ])
+
+
+def process():
+    """
+    Check whether ARM system uses path with compatible kernel/bootloader.
```
Comment (MichalHe) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:33]:
```suggestion
    Check whether the upgrade path will use a target kernel compatible with the source bootloader on ARM systems.
```

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+from leapp import reporting
+from leapp.libraries.common.config.architecture import ARCH_ARM64, matches_architecture
+from leapp.libraries.common.config.version import get_source_version, get_target_version
+from leapp.libraries.stdlib import api
+
+
+def _inhibit_upgrade():
+    title = 'Upgrade {} to {} unsupported for ARM machines.'.format(
```
Comment (MichalHe) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:8]:
```suggestion
    title = 'Upgrade from RHEL {} to RHEL {} is unsupported for ARM machines.'.format(
```

Comment (MichalHe) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:8]: Not sure whether "unsupported" is correct word - it is just not possible.  @Rezney

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+from leapp.tags import ChecksPhaseTag, IPUWorkflowTag
+
+
+class CheckArmBootloader(Actor):
+    """
+    Inhibit ARM system upgrades on path with incompatible kernel/bootloader
+
+    Due to an incompatibility of RHEL8 bootloader with newer versions of kernel
+    on RHEL9 since version 9.5 the upgrade cannot be performed as the old
+    bootloader is used to load the new versions of kernel during the upgrade.
```
Comment (MichalHe) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py:13]: Yes, but this kinda does not describes the issue. The old loader is always used to load the new kernel. Maybe rephrase it to something like `the old bootloader cannot load the new kernel`.

**Review by MichalHe — COMMENTED** [review]
Cool patch! Let's add the KCS. I will attempt to test it and approve accordingly.

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 1 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:33]: Much better wording. Sure!

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 2 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:8]: We might go with something like "impossible", "not possible", "unfeasible"  as well ...

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py (hunk 3 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py:13]: Sure. Rephrased!

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 2 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:8]: not possible it is then!

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    Check whether the upgrade path will use a target kernel compatible with the source bootloader on ARM systems
+    """
+
+    if not matches_architecture(ARCH_ARM64):
+        api.current_logger().info('Architecture not ARM. Skipping bootloader check.')
+        return
+
+    target_major, target_minor = tuple(map(int, get_target_version().split('.')))
+    if (target_major, target_minor) < (9, 5):
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:40]: * from leapp.libraries.common.config.version import matches_target_version

```suggestion
    if matches_target_version('< 9.5'):
```

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+from leapp.tags import ChecksPhaseTag, IPUWorkflowTag
+
+
+class CheckArmBootloader(Actor):
+    """
+    Inhibit ARM system upgrades on path with incompatible kernel/bootloader
+
+    Due to an incompatibility of RHEL8 bootloader with newer versions of kernel
+    on RHEL9 since version 9.5, the upgrade cannot be performed as the old
+    bootloader cannot load the new kernel when entering the interim phase.
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py:13]: It's better to make it clear that we do not expect this to be there forever

```suggestion
    bootloader cannot load the new kernel when entering the interim phase.
    
    This is temporary workaround until the issue is resolved.
```

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+from leapp import reporting
+from leapp.libraries.common.config.architecture import ARCH_ARM64, matches_architecture
+from leapp.libraries.common.config.version import get_source_version, get_target_version
+from leapp.libraries.stdlib import api
+
+
+def _inhibit_upgrade():
+    title = 'Upgrade RHEL {} to RHEL {} not possible for ARM machines.'.format(
+        get_source_version(), get_target_version())
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:9]: nitpick; you can keep the original if you want to.

Officially the upgrade to RHEL 9.5 is supported only from RHEL 8.10, so from this point I would even completely strip the info from what system they are upgrading. I was also thinking to replace target version directly by 9.5 string, but as people will start soon to test upgrade 8.10 -> 9.6, let's keep it like this until we apply the proper fix - which we know, it must be delivered for 9.6 upgrades for sure. Once the fix is applied, this one actor will be removed.

```suggestion
    title = 'Cannot upgrade to RHEL {} on ARM machines.'.format(get_target_version())
```

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    summary = ('Due to the incompatibility of the RHEL8 bootloader with a '
+               'newer version of kernel on RHEL9 a direct upgrade '
+               'cannot be performed for versions 9.5 and up.')
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:12]: This could be little bit disturbing for users as they will read it as: _upgrades to 9.5 and higher will not be possible on arm_. From this point we want to make it more clear that this is temporary solution but
* we do not want to make a promise neither - we should not regarding RH policies
* and we want to be clear that we do not expect any hotfix for RHEL 9.5 neither

```suggestion
    summary = (
        'Due to the incompatibility of the RHEL 8 bootloader with a newer version of kernel on RHEL {}'
        ' for ARM machines, the direct upgrade cannot be performed to this RHEL'
        ' system version now. The fix is not expected to be delivered during the RHEL 9.5 lifetime.'
        .format(get_target_version())
    )        
```

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        reporting.ExternalLink(
+            title='Known issues for the RHEL 8.10 to RHEL 9.5 upgrade',
+            url='https://red.ht/upgrading-rhel8-to-rhel9-known-issues'),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Groups([reporting.Groups.INHIBITOR]),
+        reporting.Groups([reporting.Groups.SANITY]),
+        reporting.Remediation(hint=(
+            'To upgrade to a RHEL version 9.5 and up, first perform an upgrade '
+            'to version 9.4 using the leapp `--target` option and then perform '
+            'a minor dnf update after the upgrade')),
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:26]:
```suggestion
            'To upgrade to the RHEL {} version, first in-place upgrade to RHEL 9.4 instead'
            ' using the leapp `--target=9.4` option. After you finish the upgrade - including'
            ' all required manual post-upgrade steps as well -'
            '  update to the newer minor version using the dnf tool. In case of using Red Hat'
            ' subscription-manager, do not forget to change the lock version to the newer one'
            ' or unset the version lock before using DNF to be able to perform the minor version update.'
            ' You can use e.g. `subscription-manager release --unset` command for that.'
            .format(get_target_version())
        )),
```

**Review by pirat89 — CHANGES_REQUESTED** [review]
Suggested some minor changes in report and one code improvement to use existing functions for checking the system versions.

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 4 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py]: Much better! I was suspicious we had to have something like that

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py (hunk 5 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/actor.py:13]: Sure! Fixed

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 6 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py:9]: I will keep it as is. When executed as supposed to nothing happens. When this somehow gets run when we don't expect it to, we get more information immediately

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 7 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py]: Yes I agree. Fixed!

File: repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py (hunk 8 above)
Comment (dkubek) [reply repos/system_upgrade/el8toel9/actors/checkarmbootloader/libraries/checkarmbootloader.py]: Thanks a lot for expanding this! Fixed

**Review by pirat89 — APPROVED** [review]
lgtm and based on the simplicity of the solution and provided tests I believe it works as expected.
